### PR TITLE
Fix: GPIO Stepper Driver

### LINKED ISSUE #80875: Stepper API: stepper.h and stepper_shell.c
**Describe the bug**

1. incorrect c-prototype and function definition [set_callback](https://github.com/zephyrproject-rtos/zephyr/blob/5c9b0a5a33cab280ed29bb73e854a355fcd4194b/include/zephyr/drivers/stepper.h#L452) in stepper.h 
2. Absence of `NULL` check in print_callback in [stepper_shell.c](https://github.com/zephyrproject-rtos/zephyr/blob/5c9b0a5a33cab280ed29bb73e854a355fcd4194b/drivers/stepper/stepper_shell.c#L46) is causing HARD_FAULT while using `stepper_shell`

**To Reproduce**

1. Visible in stepper.h
2. Discovered while using stepper shell: 
    - flash stepper_shell
    - `stepper set_max_velocity <your_stepper> 1`
    - `stepper set_target_position <your_stepper> 1`

**Expected behavior**

1. No compile time warnings of `set_callback` defined but not used.
2. shell_output: `"your_stepper: Steps completed."`

**Impact**

annoyance.

**Logs and console output**

```
uart:~$ 
[00:00:30.546,000] <err> os: ***** HARD FAULT *****
[00:00:30.546,000] <err> os: r0/a1:  0x0800b589  r1/a2:  0x00001000  r2/a3:  0x00000001
[00:00:30.546,000] <err> os: r3/a4:  0x0800b589 r12/ip:  0x00000063 r14/lr:  0x0800a003
[00:00:30.546,000] <err> os:  xpsr:  0x01000000
[00:00:30.546,000] <err> os: Faulting instruction address (r15/pc): 0x0800c2a6
[00:00:30.546,000] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
[00:00:30.546,000] <err> os: Current thread: 0x200012d0 (sysworkq)
[00:00:30.617,000] <err> os: Halting system
uart:~$ 
```

### PR DESCRIPTION
**Pull request by jilaypandya**
- Singularize instantiation of gpio-stepper driver
- Add missing SPDX headers
- Document set_event_callback function in steppers.rst
- Use the correct dtcompatible in release notes
- Rename tests from stepper_api to gpio_steppers
- Add collaborators in stepper drivers

### CODE REVIEW

File: doc/hardware/peripherals/stepper.rst
Comment (kartben) [inline doc/hardware/peripherals/stepper.rst:27]:
```suggestion
- Register an **event callback** using :c:func:`stepper_set_event_callback`.
```

File: doc/releases/release-notes-4.0.rst
Comment (kartben) [inline doc/releases/release-notes-4.0.rst:337]: fwiw the use of plural is a bit awkward and I am having a hard time finding other bindings using plural form, especially controllers

**Review by kartben — CHANGES_REQUESTED** [review]

File: doc/releases/release-notes-4.0.rst
Comment (jilaypandya) [reply doc/releases/release-notes-4.0.rst:337]: Yeah i'll simplify the driver and make it singular

File: dts/bindings/stepper/zephyr,gpio-stepper.yaml
Comment (faxe1008) [inline dts/bindings/stepper/zephyr,gpio-stepper.yaml:24]: Don't know if this is common practice for drivers generally or if it's fine that if more fields are added they simply go unused?

Comment (jilaypandya) [reply dts/bindings/stepper/zephyr,gpio-stepper.yaml]: you are right atleast in this case all the properties are used. hence have removed `property_allowlist`

File: drivers/stepper/stepper_shell.c
Comment (jilaypandya) [reply drivers/stepper/stepper_shell.c:353]: Passing now `shell_instance` along and also have included the `NULL` pointer check

Comment (dipakgmx) [reply drivers/stepper/stepper_shell.c:353]: You can stick with `NULL`. 

See [can_shell.c](https://github.com/zephyrproject-rtos/zephyr/blob/193eeaef0c04570b80a954b58ccd98075150fb11/drivers/can/can_shell.c#L926), where they the user data is left as `NULL`. (see [can_add_rx_filter()](https://docs.zephyrproject.org/apidoc/latest/group__can__interface.html#gae9dd69a13b960f773ab07bb0bb13b5e9))

File: drivers/stepper/stepper_shell.c
Comment (dipakgmx) [inline drivers/stepper/stepper_shell.c:399]: Same here. Revert to `NULL`.

File: drivers/stepper/stepper_shell.c
Comment (jilaypandya) [reply drivers/stepper/stepper_shell.c:353]: The `shell_instance` is required later in the `print_callback`. I think it's a bit different as to how it is handled in this shell as compared to `can_shell`.

Comment (dipakgmx) [reply drivers/stepper/stepper_shell.c:353]: My bad. Then it makes sense to pass the shell.

File: drivers/stepper/gpio_stepper_controller.c
Comment (bjarki-andreasen) [inline drivers/stepper/gpio_stepper_controller.c:85]: `GPIO_ACTIVE_LOW` is actually 1 `#define GPIO_ACTIVE_LOW         (1 << 0)`, its used to configure the polarity of the pin, you should simply pass 0

**Review by bjarki-andreasen — CHANGES_REQUESTED** [review]
Regarding the name change of the test from `stepper_api` to `gpio_stepper`, is it no longer possible to use that test with other drivers than `gpio_stepper`?

Regarding collaborators, you can add me as well `bjarki-andreasen` if you wish :)

**Comment by jilaypandya** [comment]
> Regarding the name change of the test from `stepper_api` to `gpio_stepper`, is it no longer possible to use that test with other drivers than `gpio_stepper`?

Yeah that's true, In the current state it can only be tested only for the gpio stepper. For other controllers an emul would have to be used in order to test throse controllers. I think the idea of creating the API test in this form was incorrect.

**Comment by bjarki-andreasen** [comment]
> > Regarding the name change of the test from `stepper_api` to `gpio_stepper`, is it no longer possible to use that test with other drivers than `gpio_stepper`?
> 
> Yeah that's true, In the current state it can only be tested only for the gpio stepper. For other controllers an emul would have to be used in order to test throse controllers. I think the idea of creating the API test in this form was incorrect.

Well, I know we can't get actual feedback from the mechanical position of the stepper, but we do get feedback in the form of `STEPPER_EVENT_STEPS_COMPLETED`, and we can determine approximately how long it should take to reach the given position. Would the test suite not run perfectly fine on real hardware? its not like we are reading back the actual position from the gpio_stepper driver either, we just wait :)

**Comment by jilaypandya** [comment]
> > > Regarding the name change of the test from `stepper_api` to `gpio_stepper`, is it no longer possible to use that test with other drivers than `gpio_stepper`?
> 
> > 
> 
> > Yeah that's true, In the current state it can only be tested only for the gpio stepper. For other controllers an emul would have to be used in order to test throse controllers. I think the idea of creating the API test in this form was incorrect.
> 
> 
> 
> Well, I know we can't get actual feedback from the mechanical position of the stepper, but we do get feedback in the form of `STEPPER_EVENT_STEPS_COMPLETED`, and we can determine approximately how long it should take to reach the given position. Would the test suite not run perfectly fine on real hardware? its not like we are reading back the actual position from the gpio_stepper driver either, we just wait :)

Yeah atleast for gpio stepper this works fine on the hardware as well, I mean since we are just toggling the gpios. And that was also the reason why I renamed it from stepper_api to gpio_stepper. For tmc5041 I have something like an emulator in mind to test it thoroughly.

But yeah this stepper_api test_suite can test only gpio stepper and cannot actually be used for other drivers.

**Comment by bjarki-andreasen** [comment]
> > > > Regarding the name change of the test from `stepper_api` to `gpio_stepper`, is it no longer possible to use that test with other drivers than `gpio_stepper`?
> > 
> > > 
> > 
> > > Yeah that's true, In the current state it can only be tested only for the gpio stepper. For other controllers an emul would have to be used in order to test throse controllers. I think the idea of creating the API test in this form was incorrect.
> > 
> > 
> > 
> > Well, I know we can't get actual feedback from the mechanical position of the stepper, but we do get feedback in the form of `STEPPER_EVENT_STEPS_COMPLETED`, and we can determine approximately how long it should take to reach the given position. Would the test suite not run perfectly fine on real hardware? its not like we are reading back the actual position from the gpio_stepper driver either, we just wait :)
> 
> Yeah atleast for gpio stepper this works fine on the hardware as well, I mean since we are just toggling the gpios. And that was also the reason why I renamed it from stepper_api to gpio_stepper. For tmc5041 I have something like an emulator in mind to test it thoroughly.
> 
> But yeah this stepper_api test_suite can test only gpio stepper and cannot actually be used for other drivers.

Even though this test suite is not fully covering the controller, it would run the exact same as the gpio driven driver though, making the test able to run on any hardware right? Just to tezt a base set of features

**Comment by jilaypandya** [comment]
> > > > > Regarding the name change of the test from `stepper_api` to `gpio_stepper`, is it no longer possible to use that test with other drivers than `gpio_stepper`?
> > > 
> > > > 
> > > 
> > > > Yeah that's true, In the current state it can only be tested only for the gpio stepper. For other controllers an emul would have to be used in order to test throse controllers. I think the idea of creating the API test in this form was incorrect.
> > > 
> > > 
> > > 
> > > Well, I know we can't get actual feedback from the mechanical position of the stepper, but we do get feedback in the form of `STEPPER_EVENT_STEPS_COMPLETED`, and we can determine approximately how long it should take to reach the given position. Would the test suite not run perfectly fine on real hardware? its not like we are reading back the actual position from the gpio_stepper driver either, we just wait :)
> > 
> > Yeah atleast for gpio stepper this works fine on the hardware as well, I mean since we are just toggling the gpios. And that was also the reason why I renamed it from stepper_api to gpio_stepper. For tmc5041 I have something like an emulator in mind to test it thoroughly.
> > 
> > But yeah this stepper_api test_suite can test only gpio stepper and cannot actually be used for other drivers.
> 
> Even though this test suite is not fully covering the controller, it would run the exact same as the gpio driven driver though, making the test able to run on any hardware right? Just to tezt a base set of features

Ahh yes, now I get what you are pointing to. Thanks 😊. I'll revert it back😊

**Review by dipakgmx — DISMISSED** [review]

**Review by faxe1008 — DISMISSED** [review]

**Review by bjarki-andreasen — DISMISSED** [review]

**Comment by dkalowsk** [comment]
That second commit is kind of doing a lot of things here.  Usually I'd flag this as the commit not "doing just one thing".  @kartben what I don't know here is if the change to the MAINTAINERs needs to go through the TSC or not.  

FYI @mmahadevan108

**Review by fabiobaltieri — DISMISSED** [review]
Yeah commit like this should be split into individual commits, pretty much one one for each point, easier to review, bisect etc... anyway I think it's fine to go with this for this time.

**Comment by mmahadevan108** [comment]
@MaureenHelm, this PR makes changes to the `MAINTAINERS` file. Can you please take a look. Do we need TSC approval to add collaborators?

**Comment by fabiobaltieri** [comment]
> Do we need TSC approval to add collaborators?

That is definitely not the case, the recent thing is to add the actual collaborator permissions to users that did not have it previously, adding collaborators is pretty inconsequential.

**Comment by mmahadevan108** [comment]
@jilaypandya , have addressed the changes requested by @kartben

**Comment by mmahadevan108** [comment]
@MaureenHelm can you help review

**Review by kartben — DISMISSED** [review]

**Comment by henrycoding** [comment]
so when will this request be merged?

**Comment by fabiobaltieri** [comment]
> so when will this request be merged?

The project is in hard freeze for the 4.0 release, feature PRs will be merged next week.

**Comment by mmahadevan108** [comment]
After discussing with @jilaypandya , changing the milestone to 4.1
@jilaypandya , please add the `backport-v4.0` label if this is needed for 4.0

**Comment by nashif** [comment]
why is this labled with TSC?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
> That second commit is kind of doing a lot of things here. Usually I'd flag this as the commit not "doing just one thing".

> Yeah commit like this should be split into individual commits, pretty much one one for each point, easier to review, bisect etc.

The MAINTAINERS commit looks fine, but the second commit needs to be split up.

**Review by dipakgmx — DISMISSED** [review]

**Review by MaureenHelm — DISMISSED** [review]

**Review by faxe1008 — DISMISSED** [review]

**Review by kartben — CHANGES_REQUESTED** [review]
release notes will need to be moved to 4.1, and migraiton guide entry added to indicate the rename of the compatible

**Comment by jilaypandya** [comment]
> release notes will need to be moved to 4.1, and migraiton guide entry added to indicate the rename of the compatible

Yep i have dropped the backport label and the release notes change. This fixes can come in with 4.1

**Review by dipakgmx — DISMISSED** [review]

**Review by bjarki-andreasen — DISMISSED** [review]

**Comment by kartben** [comment]
> > release notes will need to be moved to 4.1, and migraiton guide entry added to indicate the rename of the compatible
> 
> Yep i have dropped the backport label and the release notes change. These fixes can come in with 4.1

Migration guide entry is missing. Renaming the compatible means people will need to update their code

**Comment by jilaypandya** [comment]
> > > release notes will need to be moved to 4.1, and migraiton guide entry added to indicate the rename of the compatible
> > 
> > 
> > Yep i have dropped the backport label and the release notes change. These fixes can come in with 4.1
> 
> Migration guide entry is missing. Renaming the compatible means people will need to update their code

can it be a seperate PR?, i will reference it over here.

**Comment by jilaypandya** [comment]
> > > release notes will need to be moved to 4.1, and migraiton guide entry added to indicate the rename of the compatible
> > 
> > 
> > Yep i have dropped the backport label and the release notes change. These fixes can come in with 4.1
> 
> Migration guide entry is missing. Renaming the compatible means people will need to update their code

added

**Review by bjarki-andreasen — APPROVED** [review]

**Review by dipakgmx — APPROVED** [review]

**Review by kartben — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**
